onnxruntime/core/framework/execution_provider.cc

```diff
@@ -200,27 +200,27 @@ int IExecutionProvider::ModelMetadefIdGenerator::GenerateId(const onnxruntime::G
   } else {
     uint32_t hash[4] = {0, 0, 0, 0};
 
+    auto hash_str = [&hash](const std::string& str) {
+      MurmurHash3::x86_128(str.data(), gsl::narrow_cast<int32_t>(str.size()), hash[0], &hash);
+    };
+
     // prefer path the model was loaded from
     // this may not be available if the model was loaded from a stream or in-memory bytes
     const auto& model_path_str = main_graph.ModelPath().ToPathString();
     if (!model_path_str.empty()) {
-      MurmurHash3::x86_128(model_path_str.data(), gsl::narrow_cast<int32_t>(model_path_str.size()), hash[0], &hash);
-    } else {
-      auto hash_str = [&hash](const std::string& str) {
-        MurmurHash3::x86_128(str.data(), gsl::narrow_cast<int32_t>(str.size()), hash[0], &hash);
-      };
-
-      // fingerprint the main graph by hashing graph inputs and the ordered outputs from each node
-      for (const auto* node_arg : main_graph.GetInputsIncludingInitializers()) {
-        hash_str(node_arg->Name());
-      }
+      hash_str(model_path_str);
+    }
+
+    // fingerprint the main graph by hashing graph inputs and the ordered outputs from each node
+    for (const auto* node_arg : main_graph.GetInputsIncludingInitializers()) {
+      hash_str(node_arg->Name());
+    }
 
```

Comment:
Hashing the inputs, initializers and all node outputs is a pretty expensive way to get the hash. 

Would an alternative option be to hash the model path multiple times (e.g. keep doing it until you have hashed a certain number of chars)?

Reply:
But we seem okay with it depending on the model load api the user chooses? 
I think it's better to just consistently do the same thing for the same model.

Reply:
I understand the cost may be higher, but we do want to hash more info besides model path, because some time users fine tunes their models a little bit but still keeps the same name. In that case we need to generate a new engine with a new ID.

Reply:
Yes. I also agree to remove model path in hashing. Users usually generate the engine in local machine and deploy it to production environment. Most likely model path will change, which means the engine can't be reused if we hash the model path.

Reply:
@jywu-msft I'd prefer not to do the expensive hashing, but just doing graph input/outputs/initializers seems likely to be insufficient, especially if the graph/subgraph is small. Could maybe limit to a maximum number of nodes though. Open to suggestions for other ways to relatively cheaply get the hash for a Graph.

> Users usually generate the engine in local machine and deploy it to production environment.
@stevenlix Isn't that TRT specific? This is used by multiple other EPs so we need to also take into account their usage.

The id generation was intended for creating a unique id for a MetaDef that is valid across all levels of the graph (it is 
in a class called ModelMetadefIdGenerator after all) rather than as a way to track a particular model across different sessions or file moves. For its intended purpose the path is sufficient if available (assuming enough characters are hashed). 

If you want to force full hashing maybe an additional parameter to do that could be added and defaulted to off. 

Or could maybe separate the full hashing of a Graph into a standalone helper so you can use that directly if needed to get a value that is stable across file moves.